File: server/enrichment_service.ts
import { extractVideoId, fetchYouTubeMetadata, YouTubeMetadata } from "./youtube_service";
import { analyzeVideoMetadata, GeminiAnalysis, GeminiQuotaExhaustedError } from "./gemini_service";
import crypto from "crypto";

export interface EnrichedEvent {
  event_id: string;
  user_id: string;
  platform: string;
  enrichment_status: string;
  enrichment_version: number;
  browser_event: {
    content_title: string;
    url: string;
    timestamp_start: string;
    timestamp_end: string;
    duration_seconds: number;
    created_at: string;
  };
  youtube_metadata?: YouTubeMetadata;
  gemini_analysis?: GeminiAnalysis;
  processing_status: {
    youtube_metadata_fetched: boolean;
    gemini_processed: boolean;
    ready_for_processing: boolean;
    processed_at: string;
  };
  updated_at?: string;
}

export const CURRENT_ENRICHMENT_VERSION = 1;

export function generateEventId(eventData: any): string {
  if (eventData.event_id) {
    return eventData.event_id;
  }
  const userId = eventData.user_id || "unknown_user";
  const url = eventData.url || "";
  const tsStart = eventData.timestamp_start || "";
  const rawStr = `${userId}_${url}_${tsStart}`;
  return crypto.createHash("sha256").update(rawStr).digest("hex").slice(0, 24);
}

export async function enrichEventPipeline(eventData: any, db: any): Promise<EnrichedEvent> {
  const url = eventData.url || "";
  const contentTitle = eventData.content_title || "";
  const userId = eventData.user_id || "";
  const eventId = generateEventId(eventData);

  // 1. Extract Video ID
  const videoId = extractVideoId(url);
  if (!videoId) {
    console.warn(`[ENRICHMENT] No YouTube video ID could be extracted from: ${url}`);
    return {
      event_id: eventId,
      user_id: userId,
      platform: eventData.platform || "youtube",
      enrichment_status: "skipped",
      enrichment_version: CURRENT_ENRICHMENT_VERSION,
      browser_event: {
        content_title: contentTitle,
        url,
        timestamp_start: eventData.timestamp_start || "",
        timestamp_end: eventData.timestamp_end || "",
        duration_seconds: eventData.duration_seconds || 0,
        created_at: new Date().toISOString()
      },
      processing_status: {
        youtube_metadata_fetched: false,
        gemini_processed: false,
        ready_for_processing: false,
        processed_at: new Date().toISOString()
      }
    };
  }

  // 2. Check cache in enriched_events
  let cachedDoc: any = null;
  let cachedContentDoc: any = null;
  if (db) {
    try {
      const enrichedCollection = db.collection("enriched_events");
      cachedDoc = await enrichedCollection.findOne({ event_id: eventId });

      if (!cachedDoc || cachedDoc.enrichment_status !== "completed") {
        cachedContentDoc = await enrichedCollection.findOne({
          "youtube_metadata.video_id": videoId,
          enrichment_status: "completed",
          "processing_status.gemini_processed": true
        });
      }
    } catch (err) {
      console.error("[ENRICHMENT] Error checking cache in server database:", err);
    }
  }

  if (cachedDoc && cachedDoc.enrichment_status === "completed" && cachedDoc.enrichment_version === CURRENT_ENRICHMENT_VERSION) {
    console.log(`[ENRICHMENT] Found cached enrichment for event_id=${eventId}`);
    return cachedDoc;
  }

  let youtubeMetadata: YouTubeMetadata | undefined = undefined;
  let geminiAnalysis: GeminiAnalysis | undefined = undefined;

  if (cachedContentDoc) {
    console.log(`[ENRICHMENT] Reusing content analysis for video ${videoId}: event_id=${eventId}`);
    youtubeMetadata = cachedContentDoc.youtube_metadata;
    geminiAnalysis = cachedContentDoc.gemini_analysis;
  } else if (cachedDoc) {
    youtubeMetadata = cachedDoc.youtube_metadata;
    geminiAnalysis = cachedDoc.enrichment_status === "completed" ? cachedDoc.gemini_analysis : undefined;
  }

  // 3. If not cached, fetch metadata and analyze with Gemini
  if (!youtubeMetadata) {
    youtubeMetadata = await fetchYouTubeMetadata(videoId, contentTitle);
  }

  let quotaExhausted = false;
  if (!geminiAnalysis) {
    try {
      geminiAnalysis = await analyzeVideoMetadata(youtubeMetadata);
    } catch (err) {
      if (err instanceof GeminiQuotaExhaustedError) {
        console.warn(`[ENRICHMENT] Gemini quota exhausted for event_id=${eventId}`);
        quotaExhausted = true;
        geminiAnalysis = undefined;
      } else {
        console.error(`[ENRICHMENT] Gemini processing failed for event_id=${eventId}:`, err);
        geminiAnalysis = undefined;
      }
    }
  }

  const enrichmentStatus = geminiAnalysis ? "completed" : (quotaExhausted ? "quota_exhausted" : "failed");
  const geminiProcessed = !!geminiAnalysis;

  // 4. Form final enriched document
  const enrichedDoc: EnrichedEvent = {
    event_id: eventId,
    user_id: userId,
    platform: "youtube",
    enrichment_status: enrichmentStatus,
    enrichment_version: CURRENT_ENRICHMENT_VERSION,
    browser_event: {
      content_title: contentTitle,
      url,
      timestamp_start: eventData.timestamp_start || "",
      timestamp_end: eventData.timestamp_end || "",
      duration_seconds: eventData.duration_seconds || 0,
      created_at: new Date().toISOString()
    },
    youtube_metadata: youtubeMetadata,
    gemini_analysis: geminiAnalysis,
    processing_status: {
      youtube_metadata_fetched: youtubeMetadata ? (youtubeMetadata.official_title !== contentTitle) : false,
      gemini_processed: geminiProcessed,
      ready_for_processing: true,
      processed_at: new Date().toISOString()
    },
    updated_at: new Date().toISOString()
  };

  // 5. Save in database using UPSERT
  if (db) {
    try {
      const enrichedCollection = db.collection("enriched_events");
      await enrichedCollection.updateOne(
        { event_id: eventId },
        { $set: enrichedDoc },
        { upsert: true }
      );
      console.log(`[ENRICHMENT] Stored final enriched event (event_id=${eventId}, status=${enrichmentStatus}) for user ${userId}`);
    } catch (err) {
      console.error("[ENRICHMENT] Failed to save enriched event to database collection:", err);
    }
  }

  return enrichedDoc;
}

File: enrichment_service.py
import os
import hashlib
from datetime import datetime
from youtube_service import extract_video_id, fetch_youtube_metadata
from gemini_service import analyze_video_metadata, GeminiQuotaExhaustedError

CURRENT_ENRICHMENT_VERSION = 1

def generate_event_id(event_data: dict) -> str:
    """
    Generates a deterministic unique identifier for a specific watch event session.
    Format: SHA256(user_id_url_timestamp_start)
    """
    if event_data.get("event_id"):
        return event_data["event_id"]
    
    user_id = event_data.get("user_id", "unknown_user")
    url = event_data.get("url", "")
    ts_start = event_data.get("timestamp_start", "")
    
    raw_str = f"{user_id}_{url}_{ts_start}"
    return hashlib.sha256(raw_str.encode("utf-8")).hexdigest()[:24]

async def enrich_event_pipeline(event_data: dict, db=None) -> dict:
    url = event_data.get("url", "")
    content_title = event_data.get("content_title", "")
    user_id = event_data.get("user_id", "")
    ts_start = event_data.get("timestamp_start", "")
    ts_end = event_data.get("timestamp_end", "")
    duration = event_data.get("duration_seconds", 0)
    
    event_id = generate_event_id(event_data)
    video_id = extract_video_id(url)
    
    if not video_id:
        print(f"[ENRICHMENT] No YouTube video ID could be extracted from: {url}")
        return {
            "event_id": event_id,
            "user_id": user_id,
            "platform": event_data.get("platform", "youtube"),
            "enrichment_status": "skipped",
            "enrichment_version": CURRENT_ENRICHMENT_VERSION,
            "browser_event": {
                "content_title": content_title,
                "url": url,
                "timestamp_start": ts_start,
                "timestamp_end": ts_end,
                "duration_seconds": duration,
                "created_at": datetime.utcnow().isoformat() + "Z"
            },
            "processing_status": {
                "youtube_metadata_fetched": False,
                "gemini_processed": False,
                "ready_for_processing": False,
                "processed_at": datetime.utcnow().isoformat() + "Z"
            }
        }
    
    # 2. Check cache in enriched_events
    cached_doc = None
    cached_content_doc = None
    if db is not None:
        try:
            # Check by specific event_id first
            cached_doc = await db.enriched_events.find_one({"event_id": event_id})
            
            # Also check if another watch event for the same video has completed content analysis
            if not cached_doc or cached_doc.get("enrichment_status") != "completed":
                cached_content_doc = await db.enriched_events.find_one({
                    "youtube_metadata.video_id": video_id,
                    "enrichment_status": "completed",
                    "processing_status.gemini_processed": True
                })
        except Exception as e:
            print(f"[ENRICHMENT] Error checking cache in MongoDB: {e}")
            
    youtube_metadata = None
    gemini_analysis = None
    
    # Verify if cached_doc has completed enrichment
    if cached_doc and cached_doc.get("enrichment_status") == "completed" and cached_doc.get("enrichment_version") == CURRENT_ENRICHMENT_VERSION:
        print(f"[ENRICHMENT] Cache hit: event_id={event_id}")
        return cached_doc
    
    if cached_content_doc:
        print(f"[ENRICHMENT] Reusing cached content analysis for video {video_id}: event_id={event_id}")
        youtube_metadata = cached_content_doc.get("youtube_metadata")
        gemini_analysis = cached_content_doc.get("gemini_analysis")
    elif cached_doc:
        youtube_metadata = cached_doc.get("youtube_metadata")
        gemini_analysis = cached_doc.get("gemini_analysis") if cached_doc.get("enrichment_status") == "completed" else None

    # 3. If missing metadata, fetch YouTube metadata
    if not youtube_metadata:
        youtube_metadata = await fetch_youtube_metadata(video_id, content_title)
        
    # 4. If missing valid Gemini analysis, attempt Gemini API with bounded retries
    quota_exhausted = False
    if not gemini_analysis:
        try:
            print(f"[ENRICHMENT] Cache miss: event_id={event_id}, calling Gemini API...")
            gemini_analysis = await analyze_video_metadata(youtube_metadata, raise_on_quota=True)
        except GeminiQuotaExhaustedError as quota_err:
            print(f"[ENRICHMENT] Gemini quota exhausted: event_id={event_id}, preserving raw event for retry. Detail: {quota_err}")
            quota_exhausted = True
            gemini_analysis = None
        except Exception as gen_err:
            print(f"[ENRICHMENT] Gemini processing failed for event_id={event_id}: {gen_err}")
            gemini_analysis = None

    enrichment_status = "completed" if gemini_analysis else ("quota_exhausted" if quota_exhausted else "failed")
    gemini_processed = bool(gemini_analysis is not None)

    # 5. Form final enriched document
    enriched_doc = {
        "event_id": event_id,
        "user_id": user_id,
        "platform": "youtube",
        "enrichment_status": enrichment_status,
        "enrichment_version": CURRENT_ENRICHMENT_VERSION,
        "browser_event": {
            "content_title": content_title,
            "url": url,
            "timestamp_start": ts_start,
            "timestamp_end": ts_end,
            "duration_seconds": duration,
            "created_at": datetime.utcnow().isoformat() + "Z"
        },
        "youtube_metadata": youtube_metadata,
        "gemini_analysis": gemini_analysis,
        "processing_status": {
            "youtube_metadata_fetched": youtube_metadata.get("official_title") != content_title if youtube_metadata else False,
            "gemini_processed": gemini_processed,
            "ready_for_processing": True,
            "processed_at": datetime.utcnow().isoformat() + "Z"
        },
        "updated_at": datetime.utcnow().isoformat() + "Z"
    }
    
    # 6. Store in MongoDB via UPSERT (idempotent write by event_id)
    if db is not None:
        try:
            await db.enriched_events.update_one(
                {"event_id": event_id},
                {"$set": enriched_doc},
                upsert=True
            )
            print(f"[ENRICHMENT] Gemini {'success' if gemini_processed else 'quota exhausted'}: stored event_id={event_id} (status={enrichment_status}) for user {user_id}")
              
            # Update matching raw events with official title
            official_title = youtube_metadata.get("official_title") if youtube_metadata else None
            if official_title and official_title not in ["Unknown YouTube Video", "YouTube Video"]:
                await db.raw_events.update_many(
                    {"user_id": user_id, "url": url},
                    {"$set": {"content_title": official_title}}
                )
        except Exception as e:
            print(f"[ENRICHMENT] Failed to upsert enriched event in MongoDB: {e}")
            
    return enriched_doc
